### PR TITLE
[Merged by Bors] - feat: Dissociation of sets

### PR DESCRIPTION
**Pull request by YaelDillies**
In additive combinatorics, a pretty standard tool called Chang's lemma uses a version of linear independence and linear span with "scalars" restricted to `{-1, 0, 1}`. This PR defines those analogs, known as dissociation and span in the literature.

From LeanAPAP

---

- [x] depends on: #10551

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Review by TFBloom — APPROVED** [review]
I've reviewed this from a mathematics perspective - all definitions are correct and reflect accurately how these important concepts are used in additive combinatorics. 

Once people are happy with the Lean, this should be merged.

**Review by grhkm21 — COMMENTED** [review]
Slight mistake, I think

File: Mathlib/Combinatorics/Additive/Dissociation.lean [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    (∏ a in t, a) / ∏ a in u, a ∈ mulSpan s :=
+  mem_mulSpan.2 ⟨Set.indicator t 1 - Set.indicator u 1, fun a ↦ by
+    by_cases a ∈ t <;> by_cases a ∈ u <;> simp [*], by simp [prod_div_distrib, zpow_sub,
+      ← div_eq_mul_inv, Set.indicator, pow_ite, inter_eq_right.2, *]⟩
+
+/-- If every dissociated subset of `s` has size at most `d`, then `s` is actually generated by a
+subset of size at most `d`.
+
+This is a dissociation analog of the fact that a set whose linearly independent subspaces all have
+dimension at most `d` is of dimension at most `d` itself. -/
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/Additive/Dissociation.lean:123]:
```suggestion
This is a dissociation analog of the fact that a set whose linearly independent subsets all have
size at most `d` is of dimension at most `d` itself. -/
```

File: Mathlib/Combinatorics/Additive/Dissociation.lean [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    (∏ a in t, a) / ∏ a in u, a ∈ mulSpan s :=
+  mem_mulSpan.2 ⟨Set.indicator t 1 - Set.indicator u 1, fun a ↦ by
+    by_cases a ∈ t <;> by_cases a ∈ u <;> simp [*], by simp [prod_div_distrib, zpow_sub,
+      ← div_eq_mul_inv, Set.indicator, pow_ite, inter_eq_right.2, *]⟩
+
+/-- If every dissociated subset of `s` has size at most `d`, then `s` is actually generated by a
+subset of size at most `d`.
+
+This is a dissociation analog of the fact that a set whose linearly independent subspaces all have
+size at most `d` is of dimension at most `d` itself. -/
```
[the remark this answers is not in the document]
Comment (YaelDillies) [reply Mathlib/Combinatorics/Additive/Dissociation.lean:123]: You're right there was a typo, but it wasn't there.

The famous "linearly independent subspaces"...

**Review by Ruben-VandeVelde — APPROVED** [review]
maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**